### PR TITLE
ci: Avoid redundant nix-store invocations when `--print-all-dependencies`

### LINKED ISSUE #220: `--print-all-dependencies` is running nix-store multiple times
## Currently

image

## Ideally

All these paths can be passed to the _same_ `nix-store` invocation:

image

### PR DESCRIPTION
**Pull request by rsrohitsingh682**
Resolves #220

### CODE REVIEW

File: crates/nix_rs/src/store.rs [hunk 1]
```diff
@@ -106,19 +104,20 @@ impl NixStoreCmd {
...
     }
 
-    /// Given a derivation path, this function recursively queries and return all
+    /// Given the derivation paths, this function recursively queries and return all
     /// of its dependencies in the Nix store.
     pub async fn nix_store_query_requisites_with_outputs(
         &self,
-        drv_path: PathBuf,
-    ) -> Result<Vec<StorePath>, NixStoreCmdError> {
+        drv_paths: HashSet<PathBuf>,
```
Comment (srid) [inline crates/nix_rs/src/store.rs:111]: This should take a slice (`&[Pathbuf]`). There's not reason for it to require a HashSet. A convenience of a single caller should not dictate API of some core function.

File: crates/nix_rs/src/store.rs [hunk 2]
```diff
@@ -106,19 +104,20 @@ impl NixStoreCmd {
...
-        cmd.args([
-            "--query",
-            "--requisites",
-            "--include-outputs",
-            drv_path.to_string_lossy().as_ref(),
-        ]);
+        let mut args = vec!["--query", "--requisites", "--include-outputs"];
+        let paths: Vec<&str> = drv_paths
+            .iter()
+            .map(|path| path.as_path().to_str().unwrap_or_default())
```
Comment (srid) [inline crates/nix_rs/src/store.rs:117]: @rsrohitsingh682 Is this code behaviourally correct? What do you think happens in the default (`unwrap_or_default`) case?

Comment (srid) [reply crates/nix_rs/src/store.rs:117]: Also, you want to understand how the `args` method works.  What does `AsRef` do? How do you pass it a slice of `PathBuf`?

Comment (srid) [reply crates/nix_rs/src/store.rs:117]: I see the same code in #212. We should not reflexively use Default instance without understanding what it does.

**Comment by rsrohitsingh682** [comment]
Converting this to draft as I would like to do similar changes to `nix_store_query_requisites_with_outputs` function as well.

File: crates/nix_rs/src/store.rs [hunk 3]
```diff
@@ -147,38 +147,41 @@ impl NixStoreCmd {
...
         crate::command::trace_cmd(&cmd);
         let out = cmd.output().await?;
         if out.status.success() {
-            let drv_path = String::from_utf8(out.stdout)?.trim().to_string();
-            if drv_path == "unknown-deriver" {
+            let out = String::from_utf8(out.stdout)?.trim().to_string();
+            let drv_paths: HashSet<PathBuf> = out
+                .lines()
+                .map(|l| l.trim().to_string())
+                .filter(|l| !l.is_empty())
```
Comment (srid) [inline crates/nix_rs/src/store.rs:178]: Why do we need these two lines? What happens if you get rid of them?

File: crates/nix_rs/src/store.rs [hunk 4]
```diff
@@ -188,19 +191,15 @@ impl NixStoreCmd {
...
 
-    /// Given a derivation path, this function recursively queries and return all
+    /// Given the derivation paths, this function recursively queries and return all
     /// of its dependencies in the Nix store.
     pub async fn nix_store_query_requisites_with_outputs(
         &self,
-        drv_path: PathBuf,
-    ) -> Result<Vec<StorePath>, NixStoreCmdError> {
+        drv_paths: &[PathBuf],
+    ) -> Result<HashSet<StorePath>, NixStoreCmdError> {
```
Comment (srid) [inline crates/nix_rs/src/store.rs:199]: This should continue to return `Vec` (so as to maintain original order).

Also, I see no reason for it to require a HashSet. The convenience of a single caller should not universally dictate API of some core function.

File: crates/nix_rs/src/store.rs [hunk 5]
```diff
@@ -147,38 +147,41 @@ impl NixStoreCmd {
...
     }
 
-    /// Return the derivation used to build the given build output.
+    /// Return the derivations used to build the given build output.
     pub async fn nix_store_query_deriver(
         &self,
-        out_path: StorePath,
-    ) -> Result<PathBuf, NixStoreCmdError> {
+        out_paths: &[StorePath],
+    ) -> Result<HashSet<PathBuf>, NixStoreCmdError> {
```
Comment (srid) [inline crates/nix_rs/src/store.rs:163]: This should return `Vec` (so as to maintain original order).

Also, I see no reason for it to require a HashSet over Vec. The convenience of a single caller should not universally dictate API of some core function.

File: crates/nix_rs/src/store.rs [hunk 6]
```diff
@@ -147,38 +147,41 @@ impl NixStoreCmd {
...
-        cmd.args([
-            "--query",
-            "--valid-derivers",
-            out_path.as_path().to_string_lossy().as_ref(),
-        ]);
+        let mut paths = vec![];
+        for path in out_paths {
+            paths.push(path.as_path());
+        }
+        cmd.args(["--query", "--valid-derivers"]).args(paths);
```
Comment (srid) [inline crates/nix_rs/src/store.rs:169]: Cannot this be simplified as:

```rust
cmd.args(["--query", "--valid-derivers"])
   .args(out_paths.iter().map(StorePath::as_path));
```

File: crates/nix_rs/src/store.rs (hunk 5 above)
Comment (rsrohitsingh682) [reply crates/nix_rs/src/store.rs:163]: Even I thought about this but then chose HashSet thinking about consistency as main function `fetch_all_deps` also returns a HashSet and the other reason was HashSet would contain unique while Vec won't.

Comment (rsrohitsingh682) [reply crates/nix_rs/src/store.rs:163]: But thinking about this `The convenience of a single caller should not universally dictate API of some core function.` now I am thinking I can collect the final result in a HashSet and return.

File: crates/nix_rs/src/store.rs (hunk 3 above)
Comment (rsrohitsingh682) [reply crates/nix_rs/src/store.rs:178]: You're right. These are actually not needed as we have applied trim() before to the whole String so chances of getting whitespace or empty lines won't be there. I can safely remove these two extra lines.

Comment (rsrohitsingh682) [reply crates/nix_rs/src/store.rs:178]: I tested it, it behaves normally.

File: crates/nix_rs/src/store.rs [hunk 7]
```diff
@@ -147,38 +147,33 @@ impl NixStoreCmd {
...
+        cmd.args(["--query", "--valid-derivers"])
+            .args(out_paths.iter().map(StorePath::as_path));
+
         crate::command::trace_cmd(&cmd);
         let out = cmd.output().await?;
         if out.status.success() {
-            let drv_path = String::from_utf8(out.stdout)?.trim().to_string();
-            if drv_path == "unknown-deriver" {
+            let out = String::from_utf8(out.stdout)?.trim().to_string();
+            let drv_paths: HashSet<PathBuf> = out.lines().map(PathBuf::from).collect();
```
Comment (rsrohitsingh682) [inline crates/nix_rs/src/store.rs:172]: Converting it to HashSet first and later to Vec while returning  as I want this check for `unknown-deriver` in constant time.

**Review by srid — APPROVED** [review]
Good, for future - remember to update CHANGELOG and release history.md

**Pull request merged**